File: Player/src/main/java/ru/testsimpleapps/coloraudioplayer/managers/explorer/MediaExplorerManager.java
package ru.testsimpleapps.coloraudioplayer.managers.explorer;


import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.os.AsyncTask;
import android.provider.MediaStore;
import android.support.annotation.NonNull;
import android.util.Log;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

import ru.testsimpleapps.coloraudioplayer.App;
import ru.testsimpleapps.coloraudioplayer.managers.explorer.Data.FolderData;
import ru.testsimpleapps.coloraudioplayer.managers.explorer.Data.ItemData;
import ru.testsimpleapps.coloraudioplayer.managers.tools.FileTool;
import ru.testsimpleapps.coloraudioplayer.managers.tools.MathTool;
import ru.testsimpleapps.coloraudioplayer.managers.tools.TextTool;


public class MediaExplorerManager {

    public static final String TAG = MediaExplorerManager.class.getSimpleName();
    private static MediaExplorerManager sMediaExplorerManager;

    public interface OnDataReady {
        void onSuccess();
        void onError();
    }

    public static MediaExplorerManager getInstance() {
        if (sMediaExplorerManager == null) {
            sMediaExplorerManager = new MediaExplorerManager(App.getContext());
        }

        return sMediaExplorerManager;
    }

    /*
    * Common file constant
    * */
    private final static String MEDIA_ID = MediaStore.Audio.Media._ID;
    private final static String MEDIA_TITLE = MediaStore.Audio.Media.TITLE;
    private final static String MEDIA_PATH = MediaStore.Audio.Media.DATA;
    private final static String MEDIA_ALBUMS = MediaStore.Audio.Media.ALBUM;
    private final static String MEDIA_ARTIST = MediaStore.Audio.Media.ARTIST;
    private final static String MEDIA_DURATION = MediaStore.Audio.Media.DURATION;
    private final static String MEDIA_DATE_MODIFIED = MediaStore.Audio.Media.DATE_MODIFIED;
    private final static String MEDIA_DATE_ADDED = MediaStore.Audio.Media.DATE_ADDED;
    private final static String MEDIA_SIZE = MediaStore.Audio.Media.SIZE;

    private final static Uri MEDIA_USER = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI;
    private final static String MEDIA_TYPE = MediaStore.Audio.Media.IS_MUSIC + " != 0";
    private final static String MEDIA_NAME_ORDER = MediaStore.Audio.Media.TITLE + " ASC";

    /*
    * Objects for work
    * */
    private WeakReference<OnDataReady> mOnDataReadyWeakReference;
    private AsyncFind mAsyncFind;
    private volatile boolean mIsProcessing = false;

    private final Context mContext;
    private List<FolderData> mFoldersList;
    private List<FolderData> mArtistList;
    private List<FolderData> mAlbumsList;

    private long mTotalTime;
    private long mTotalTracks;

    private MediaExplorerManager(Context context) {
        mContext = context;
        mFoldersList = new ArrayList<>();
        mArtistList = new ArrayList<>();
        mAlbumsList = new ArrayList<>();
    }

    public synchronized boolean findMedia() {
        Cursor cursor = null;
        try {
            // Get user media files
            cursor = mContext.getContentResolver().query(MEDIA_USER,
                    new String[]{MEDIA_ID,
                            MEDIA_TITLE,
                            MEDIA_PATH,
                            MEDIA_ARTIST,
                            MEDIA_ALBUMS,
                            MEDIA_DURATION,
                            MEDIA_DATE_MODIFIED,
                            MEDIA_DATE_ADDED,
                            MEDIA_SIZE},
                    MEDIA_TYPE,
                    null,
                    MEDIA_NAME_ORDER);

            // Get media data info
            mArtistList.clear();
            mAlbumsList.clear();
            mFoldersList.clear();
            if (cursor != null && cursor.getCount() > 0) {
                while (cursor.moveToNext()) {
                    try {
                        final long id = cursor.getLong(cursor.getColumnIndex(MEDIA_ID));
                        final String path = cursor.getString(cursor.getColumnIndex(MEDIA_PATH));
                        String folder = FileTool.getFolderName(path);
                        if (TextTool.isBadName(folder)) {
                            folder = FileTool.getComplexName(path);
                        }

                        final String name = cursor.getString(cursor.getColumnIndex(MEDIA_TITLE));
                        final String artist = cursor.getString(cursor.getColumnIndex(MEDIA_ARTIST));
                        final String album = cursor.getString(cursor.getColumnIndex(MEDIA_ALBUMS));
                        final long duration = cursor.getLong(cursor.getColumnIndex(MEDIA_DURATION));
                        final long dateModified = cursor.getLong(cursor.getColumnIndex(MEDIA_DATE_MODIFIED));
                        final long dateAdded = cursor.getLong(cursor.getColumnIndex(MEDIA_DATE_ADDED));
                        final long size = cursor.getLong(cursor.getColumnIndex(MEDIA_SIZE));
                        final long bitrate = MathTool.roundToTenths(size * 8 / duration);

                        final ItemData dataItem = new ItemData(path, name, folder, album, artist, id,
                                duration, dateModified, dateAdded, bitrate);

                        // For artist grouping
                        FolderData artistsContainer = findFolderByName(mArtistList, artist);
                        if (artistsContainer == null) {
                            artistsContainer = new FolderData(artist);
                            mArtistList.add(artistsContainer);
                        }
                        artistsContainer.addItem(dataItem);

                        // For albums grouping
                        FolderData albumContainer = findFolderByName(mAlbumsList, album);
                        if (albumContainer == null) {
                            albumContainer = new FolderData(album);
                            mAlbumsList.add(albumContainer);
                        }
                        albumContainer.addItem(dataItem);


                        // For folder grouping
                        FolderData folderContainer = findFolderByName(mFoldersList, folder);
                        if (folderContainer == null) {
                            folderContainer = new FolderData(folder);
                            mFoldersList.add(folderContainer);
                        }
                        folderContainer.addItem(dataItem);

                        mTotalTime += duration;
                        mTotalTracks++;
                    } catch (RuntimeException e) {
                        Log.e(TAG, e.getMessage());
                    }
                }
            }
        } catch (RuntimeException e) {
            Log.e(TAG, e.getMessage());
            return false;
        } finally {
            if (cursor != null && !cursor.isClosed())
                cursor.close();
        }
        return true;
    }

    private FolderData findFolderByName(final List<FolderData> folderData, final String name) {
        for (FolderData item : folderData) {
            if (item.getName().equalsIgnoreCase(name)) {
                return item;
            }
        }
        return null;
    }

    synchronized public void findMediaAsync() {
        if (mAsyncFind != null && !mAsyncFind.isCancelled()) {
            mAsyncFind.cancel(true);
        }

        mAsyncFind = new AsyncFind();
        mAsyncFind.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
    }

    synchronized public void setFindCallback(@NonNull final OnDataReady onDataReady) {
        mOnDataReadyWeakReference = new WeakReference<>(onDataReady);
    }

    synchronized public void removeFindCallback() {
        mOnDataReadyWeakReference = null;
    }

    synchronized public boolean isProcessing() {
        return mIsProcessing;
    }

    private class AsyncFind extends AsyncTask<Void, Void, Boolean> {

        @Override
        protected void onPreExecute() {
            super.onPreExecute();
            mIsProcessing = true;
        }

        @Override
        protected Boolean doInBackground(Void[] objects) {
            return findMedia();
        }

        @Override
        protected void onPostExecute(Boolean isSuccess) {
            super.onPostExecute(isSuccess);
            mIsProcessing = false;
            final WeakReference<OnDataReady> onDataReadyWeakReference = mOnDataReadyWeakReference;
            if (onDataReadyWeakReference != null) {
                final OnDataReady onDataReady = onDataReadyWeakReference.get();
                if (onDataReady != null) {
                    if (isSuccess.booleanValue()) {
                        onDataReady.onSuccess();
                    } else {
                        onDataReady.onError();
                    }
                }
            }
        }

    }

    synchronized public List<FolderData> getFolders() {
        return mFoldersList;
    }

    synchronized public List<FolderData> getArtists() {
        return mArtistList;
    }

    synchronized public List<FolderData> getAlbums() {
        return mAlbumsList;
    }

    synchronized public long getTotalTracks() {
        return mTotalTracks;
    }

    synchronized public long getTotalTime() {
        return mTotalTime;
    }

}

File: player/src/main/java/ru/testsimpleapps/coloraudioplayer/managers/explorer/MediaExplorerManager.kt
package ru.testsimpleapps.coloraudioplayer.managers.explorer


import android.content.Context
import android.database.Cursor
import android.net.Uri
import android.os.AsyncTask
import android.provider.MediaStore
import android.util.Log

import java.lang.ref.WeakReference
import java.util.ArrayList

import ru.testsimpleapps.coloraudioplayer.app.App
import ru.testsimpleapps.coloraudioplayer.managers.explorer.Data.FolderData
import ru.testsimpleapps.coloraudioplayer.managers.explorer.Data.ItemData
import ru.testsimpleapps.coloraudioplayer.managers.tools.FileTool
import ru.testsimpleapps.coloraudioplayer.managers.tools.MathTool
import ru.testsimpleapps.coloraudioplayer.managers.tools.TextTool


class MediaExplorerManager private constructor(private val mContext: Context) {

    /*
    * Objects for work
    * */
    private var mOnDataReadyWeakReference: WeakReference<OnDataReady>? = null
    private var mAsyncFind: AsyncFind? = null
    @Volatile
    @get:Synchronized
    var isProcessing = false
        private set
    private val mFoldersList: MutableList<FolderData>
    private val mArtistList: MutableList<FolderData>
    private val mAlbumsList: MutableList<FolderData>

    @get:Synchronized
    var totalTime: Long = 0
        private set
    @get:Synchronized
    var totalTracks: Long = 0
        private set

    val folders: List<FolderData>
        @Synchronized get() = mFoldersList

    val artists: List<FolderData>
        @Synchronized get() = mArtistList

    val albums: List<FolderData>
        @Synchronized get() = mAlbumsList

    interface OnDataReady {
        fun onSuccess()
        fun onError()
    }

    init {
        mFoldersList = ArrayList()
        mArtistList = ArrayList()
        mAlbumsList = ArrayList()
    }

    @Synchronized
    fun findMedia(): Boolean {
        var cursor: Cursor? = null
        try {
            // Get user media files
            cursor = mContext.contentResolver.query(MEDIA_USER,
                    arrayOf(MEDIA_ID, MEDIA_TITLE, MEDIA_PATH, MEDIA_ARTIST, MEDIA_ALBUMS, MEDIA_DURATION, MEDIA_DATE_MODIFIED, MEDIA_DATE_ADDED, MEDIA_SIZE),
                    MEDIA_TYPE, null,
                    MEDIA_NAME_ORDER)

            // Get media data info
            mArtistList.clear()
            mAlbumsList.clear()
            mFoldersList.clear()
            if (cursor != null && cursor.count > 0) {
                while (cursor.moveToNext()) {
                    try {
                        val id = cursor.getLong(cursor.getColumnIndex(MEDIA_ID))
                        val path = cursor.getString(cursor.getColumnIndex(MEDIA_PATH))
                        var folder = FileTool.getFolderName(path)
                        if (TextTool.isBadName(folder!!)) {
                            folder = FileTool.getComplexName(path)
                        }

                        val name = cursor.getString(cursor.getColumnIndex(MEDIA_TITLE))
                        val artist = cursor.getString(cursor.getColumnIndex(MEDIA_ARTIST))
                        val album = cursor.getString(cursor.getColumnIndex(MEDIA_ALBUMS))
                        val duration = cursor.getLong(cursor.getColumnIndex(MEDIA_DURATION))
                        val dateModified = cursor.getLong(cursor.getColumnIndex(MEDIA_DATE_MODIFIED))
                        val dateAdded = cursor.getLong(cursor.getColumnIndex(MEDIA_DATE_ADDED))
                        val size = cursor.getLong(cursor.getColumnIndex(MEDIA_SIZE))
                        val bitrate = MathTool.roundToTenths(size * 8 / duration)

                        val dataItem = ItemData(path, name, folder, album, artist, id,
                                duration, dateModified, dateAdded, bitrate)

                        // For artist grouping
                        var artistsContainer = findFolderByName(mArtistList, artist)
                        if (artistsContainer == null) {
                            artistsContainer = FolderData(artist)
                            mArtistList.add(artistsContainer)
                        }
                        artistsContainer.addItem(dataItem)

                        // For albums grouping
                        var albumContainer = findFolderByName(mAlbumsList, album)
                        if (albumContainer == null) {
                            albumContainer = FolderData(album)
                            mAlbumsList.add(albumContainer)
                        }
                        albumContainer.addItem(dataItem)


                        // For folder grouping
                        var folderContainer = findFolderByName(mFoldersList, folder)
                        if (folderContainer == null) {
                            folderContainer = FolderData(folder)
                            mFoldersList.add(folderContainer)
                        }
                        folderContainer.addItem(dataItem)

                        totalTime += duration
                        totalTracks++
                    } catch (e: RuntimeException) {
                        Log.e(TAG, e.message)
                    }

                }
            }
        } catch (e: RuntimeException) {
            Log.e(TAG, e.message)
            return false
        } finally {
            if (cursor != null && !cursor.isClosed)
                cursor.close()
        }
        return true
    }

    private fun findFolderByName(folderData: List<FolderData>, name: String?): FolderData? {
        for (item in folderData) {
            if (item.name!!.equals(name!!, ignoreCase = true)) {
                return item
            }
        }
        return null
    }

    @Synchronized
    fun findMediaAsync() {
        if (mAsyncFind != null && !mAsyncFind!!.isCancelled) {
            mAsyncFind!!.cancel(true)
        }

        mAsyncFind = AsyncFind()
        mAsyncFind!!.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR)
    }

    @Synchronized
    fun setFindCallback(onDataReady: OnDataReady) {
        mOnDataReadyWeakReference = WeakReference(onDataReady)
    }

    @Synchronized
    fun removeFindCallback() {
        mOnDataReadyWeakReference = null
    }

    private inner class AsyncFind : AsyncTask<Void, Void, Boolean>() {

        override fun onPreExecute() {
            super.onPreExecute()
            isProcessing = true
        }

        override fun doInBackground(objects: Array<Void>): Boolean? {
            return findMedia()
        }

        override fun onPostExecute(isSuccess: Boolean) {
            super.onPostExecute(isSuccess)
            isProcessing = false
            val onDataReadyWeakReference = mOnDataReadyWeakReference
            if (onDataReadyWeakReference != null) {
                val onDataReady = onDataReadyWeakReference.get()
                if (onDataReady != null) {
                    if (isSuccess) {
                        onDataReady.onSuccess()
                    } else {
                        onDataReady.onError()
                    }
                }
            }
        }

    }

    companion object {

        val TAG = MediaExplorerManager::class.java.simpleName
        private var sMediaExplorerManager: MediaExplorerManager? = null

        val instance: MediaExplorerManager
            get() {
                if (sMediaExplorerManager == null) {
                    sMediaExplorerManager = MediaExplorerManager(App.instance)
                }

                return sMediaExplorerManager!!
            }

        /*
    * Common file constant
    * */
        private val MEDIA_ID = MediaStore.Audio.Media._ID
        private val MEDIA_TITLE = MediaStore.Audio.Media.TITLE
        private val MEDIA_PATH = MediaStore.Audio.Media.DATA
        private val MEDIA_ALBUMS = MediaStore.Audio.Media.ALBUM
        private val MEDIA_ARTIST = MediaStore.Audio.Media.ARTIST
        private val MEDIA_DURATION = MediaStore.Audio.Media.DURATION
        private val MEDIA_DATE_MODIFIED = MediaStore.Audio.Media.DATE_MODIFIED
        private val MEDIA_DATE_ADDED = MediaStore.Audio.Media.DATE_ADDED
        private val MEDIA_SIZE = MediaStore.Audio.Media.SIZE

        private val MEDIA_USER = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI
        private val MEDIA_TYPE = MediaStore.Audio.Media.IS_MUSIC + " != 0"
        private val MEDIA_NAME_ORDER = MediaStore.Audio.Media.TITLE + " ASC"
    }

}
